src/base/bittorrent/filterparserthread.cpp

```diff
@@ -148,293 +208,111 @@ int FilterParserThread::parseDATFilterFile()
-                {
-                start = endOfLine;
+            if (line == "\n"sv)
                 continue;
-            }
-
-            // Each line should follow this format:
-            // 001.009.096.105 - 001.009.096.105 , 000 , Some organization
-            // The 3rd entry is access level and if above 127 the IP range isn't blocked.
-            const int firstComma = findAndNullDelimiter(buffer.data(), ',', start, endOfLine);
-            if (firstComma != -1)
-                findAndNullDelimiter(buffer.data(), ',', firstComma + 1, endOfLine);
-
-            // Check if there is an access value (apparently not mandatory)
-            if (firstComma != -1)
-            {
-                // There is possibly one
-                const long int nbAccess = strtol(buffer.data() + firstComma + 1, nullptr, 10);
-                // Ignoring this rule because access value is too high
-                if (nbAccess > 127L)
-                {
-                    start = endOfLine;
-                    continue;
-                }
-            }
-
-            // IP Range should be split by a dash
-            const int endOfIPRange = ((firstComma == -1) ? (endOfLine - 1) : (firstComma - 1));
-            const int delimIP = findAndNullDelimiter(buffer.data(), '-', start, endOfIPRange);
-            if (delimIP == -1)
-            {
-                ++parseErrorCount;
-                addLog(tr("IP filter line %1 is malformed.").arg(nbLine));
-                start = endOfLine;
-                continue;
-            }
-
-            lt::address startAddr;
-            int newStart = trim(buffer.data(), start, delimIP - 1);
-            if (!parseIPAddress(buffer.data() + newStart, startAddr))
-            {
-                ++parseErrorCount;
-                addLog(tr("IP filter line %1 is malformed. Start IP of the range is malformed.").arg(nbLine));
-                start = endOfLine;
-                continue;
-            }
 
-            lt::address endAddr;
-            newStart = trim(buffer.data(), delimIP + 1, endOfIPRange);
-            if (!parseIPAddress(buffer.data() + newStart, endAddr))
-            {
-                ++parseErrorCount;
-                addLog(tr("IP filter line %1 is malformed. End IP of the range is malformed.").arg(nbLine));
-                start = endOfLine;
+            if ((line.find('#') == 0) || (line.find("//") == 0))
                 continue;
-            }
 
-            if ((startAddr.is_v4() != endAddr.is_v4())
-                || (startAddr.is_v6() != endAddr.is_v6()))
-                {
-                ++parseErrorCount;
-                addLog(tr("IP filter line %1 is malformed. One IP is IPv4 and the other is IPv6!").arg(nbLine));
-                start = endOfLine;
-                continue;
-            }
-
-            start = endOfLine;
+            const auto& [range, accessLevel] = parseLine<isDat>(line);
```

Comment:
I hope we can write out the exact type instead of `auto`.

Reply:
>I hope we can write out the exact type instead of `auto`...

Sorry, cannot exactly understand what do you want. Do you suggest to use older approach (explicitly declare both variables and use `std::tie`)?
[Structured binding declaration](https://en.cppreference.com/w/cpp/language/structured_binding) is one of the unquestionable advantages of C++17 for me.

>Also it should be `const auto &[range, accessLevel]`

:+1: